operator/controllers/resources/role_controller.go

```diff
@@ -74,42 +101,81 @@ func (r *RoleReconciler) Delete(ctx context.Context, obj kclient.Object) error {
 }
 
 func (r *RoleReconciler) Upsert(ctx context.Context, obj kclient.Object) error {
-	k8sResource, ok := obj.(*resourcesv5.TeleportRole)
+	// We receive an unstructured object. We convert it to a typed TeleportRole object and gracefully handle errors.
+	u, ok := obj.(*unstructured.Unstructured)
 	if !ok {
 		return fmt.Errorf("failed to convert Object into resource object: %T", obj)
 	}
+	k8sResource := &resourcesv5.TeleportRole{}
+
+	// If an error happens we want to put it in status.conditions before returning.
+	err := runtime.DefaultUnstructuredConverter.FromUnstructuredWithValidation(
+		u.Object,
+		k8sResource, true, /* returnUnknownFields */
+	)
+	newStructureCondition := getStructureConditionFromError(err)
+	meta.SetStatusCondition(&k8sResource.Status.Conditions, newStructureCondition)
+	if err != nil {
+		// We update the status conditions on exit
+		statusErr := r.Status().Update(ctx, k8sResource)
+		if statusErr != nil && !errors.Is(statusErr, &json.UnmarshalTypeError{}) {
+			log.Errorf("Failed to report error in status conditions: %s", statusErr)
+		}
```

Comment:
Suggestion:

Extract:

```go
	silentUpdateStatus := func() {
		if err := r.Status().Update(ctx, k8sResource); err != nil {
			log.Errorf("Failed to report error in status conditions: %s", err)
		}
	}
```

and then changes calls to:

```suggestion
		silentUpdateStatus()
```

Note that I dropped the special handling of err; unless we expect UnmarshalTypeError to be _very_ common then we should be OK logging them.

Reply:
Done